source/as-promise/index.ts

```diff
@@ -0,0 +1,244 @@
+import {EventEmitter} from 'events';
+import getStream = require('get-stream');
+import PCancelable = require('p-cancelable');
+import calculateRetryDelay from './calculate-retry-delay';
+import {
+	NormalizedOptions,
+	CancelableRequest,
+	Response,
+	RequestError,
+	HTTPError,
+	ReadError,
+	ParseError
+} from './types';
+import PromisableRequest, {parseBody} from './core';
+import proxyEvents from '../core/utils/proxy-events';
+
+const proxiedRequestEvents = [
+	'request',
+	'response',
+	'redirect',
+	'uploadProgress',
+	'downloadProgress'
+];
+
+export default function asPromise<T>(options: NormalizedOptions): CancelableRequest<T> {
+	let retryCount = 0;
+	let body: Buffer;
+	const emitter = new EventEmitter();
+
+	const promise = new PCancelable<T>((resolve, reject, onCancel) => {
+		const makeRequest = (): void => {
+			if (options.responseType === 'json' && options.headers.accept === undefined) {
+				options.headers.accept = 'application/json';
+			}
+
+			// Support retries
+			const {throwHttpErrors} = options;
```

Comment:
`Request` needs to throw so https://github.com/sindresorhus/got/pull/1051/files#diff-6be2971b2bb8dbf48d15ff680dd898b0R132 gets called.